firmware/ethereum.c

```diff
@@ -0,0 +1,449 @@
+		_value[6] = '0' + (num[6] / 100) % 10;
+		_value[7] = '0' + (num[6] / 10) % 10;
+		_value[8] = '0' + (num[6]) % 10;
+		_value[9] = '.';
+		_value[10] = '0' + (num[5] / 100) % 10;
+		_value[11] = '0' + (num[5] / 10) % 10;
+		_value[12] = '0' + (num[5]) % 10;
+		_value[13] = '0' + (num[4] / 100) % 10;
+		_value[14] = '0' + (num[4] / 10) % 10;
+		_value[15] = '0' + (num[4]) % 10;
+		_value[16] = '0' + (num[3] / 100) % 10;
+		_value[17] = '0' + (num[3] / 10) % 10;
+		_value[18] = '0' + (num[3]) % 10;
+		strlcpy(_value + 19, " ETH", sizeof(_value) - 19);
+	} else {
+		// value is bigger than 1e9 ETH => won't fit on display (probably won't happen unless you are Vitalik)
+		strlcpy(_value, "more than a billion ETH", sizeof(_value));
+	}
+
+	value_ptr = _value;
+	while (*value_ptr == '0' && *(value_ptr + 1) >= '0' && *(value_ptr + 1) <= '9') { // skip leading zeroes
+		value_ptr++;
+	}
+
+	static char _to1[17] = {0};
+	static char _to2[17] = {0};
+	static char _to3[17] = {0};
+
+	if (to) {
+		strcpy(_to1, "to ");
+		data2hex(to, 6, _to1 + 3);
+		data2hex(to + 6, 7, _to2);
+		data2hex(to + 13, 7, _to3);
+		_to3[14] = '?'; _to3[15] = 0;
+	} else {
+		strlcpy(_to1, "to no recipient?", sizeof(_to1));
+		strlcpy(_to2, "", sizeof(_to2));
+		strlcpy(_to3, "", sizeof(_to3));
+	}
+
+	layoutDialogSwipe(DIALOG_ICON_QUESTION,
+		"Cancel",
+		"Confirm",
+		NULL,
+		"Really send",
+		value_ptr,
+		_to1,
+		_to2,
+		_to3,
+		NULL
+	);
+}
+
+/*
+ * RLP fields:
+ * - nonce (0 .. 32)
+ * - gas_price (0 .. 32)
+ * - gas_limit (0 .. 32)
+ * - to (0, 20)
+ * - value (0 .. 32)
+ * - data (0 ..)
+ */
+
+static bool ethereum_signing_check(EthereumSignTx *msg)
+{
+	// determine if address == 0
+	bool address_zero = msg->has_to;
+	if (address_zero) {
+		for (size_t i = 0; i < msg->to.size; i++) {
+			if (msg->to.bytes[i] > 0) {
+				address_zero = false;
+				break;
+			}
+		}
+	}
+
+	// sending value to address 0
+	if (address_zero && msg->has_value && msg->value.size) {
+		return false;
+	}
+	// sending transaction to address 0 without a data field
+	if (address_zero && (!msg->has_data_length || msg->data_length == 0)) {
+		return false;
+	}
+
+	return true;
+}
+
+void ethereum_signing_init(EthereumSignTx *msg, const HDNode *node)
+{
+	signing = true;
+	sha3_256_Init(&keccak_ctx);
+
+	memset(&resp, 0, sizeof(EthereumTxRequest));
+
+	if (msg->has_data_length) {
+		if (msg->data_length == 0) {
+			fsm_sendFailure(FailureType_Failure_Other, "Invalid data length provided");
+			ethereum_signing_abort();
+			return;
+		}
+		if (!msg->has_data_initial_chunk || msg->data_initial_chunk.size == 0) {
+			fsm_sendFailure(FailureType_Failure_Other, "Data length provided, but no initial chunk");
+			ethereum_signing_abort();
+			return;
+		}
+		if (msg->data_initial_chunk.size > msg->data_length) {
+			fsm_sendFailure(FailureType_Failure_Other, "Invalid size of initial chunk");
+			ethereum_signing_abort();
+			return;
+		}
+		data_total = msg->data_length;
+	} else {
+		data_total = 0;
```

Comment:
Either check that has_data_initial_chunk is empty, or set data_total to its length (depending on what protocol you prefer).